src/test/scala/mesosphere/marathon/core/deployment/impl/DeploymentActorTest.scala

```diff
@@ -47,6 +48,24 @@ class DeploymentActorTest extends AkkaUnitTest with GroupCreation {
     config.killBatchSize returns 100
     config.killBatchCycle returns 10.seconds
 
+    tracker.setGoal(any, any, any) answers { args =>
```

Comment:
I'm not a fan of implementing logic in a mock. What do you think about
```scala
tracker.setGoal(any, any, any) returns Future.sucessful(Done)
...
verify(tracker).setGoal(instance4_1.instanceId, Goal.Decommissioned, GoalChangeReason.DeletingApp)
...
events.foreach(system.eventStream.publish)
```

This makes more sense to me. It also makes clear what needs to happen.

Reply:
I am not big fan of that either but I would still rather have it in one place inside that mock than spread around all test code. The client code is not publishing those events either, it's expecting them to be published by the instanceTracker call and since test code should imitate the real client interactions, I would prefer to stay with this implemenation.

One another approach that we use e.g. for KillService is to implement a full featured mock (see `KillServiceMock`) and use that one. I found the approach I took simpler and more transparent (it's isolated inside the test class)

Reply:
It seems like we stub out `tracker.setGoal` twice?

I also would be inclined to oppose the complex mock. Have we considered just instantiating an InstanceTracker backed by an in-memory store? Judging by the argument list needed to instantiate one, it seems doable. I've done that in several areas and it runs fast enough, and removes the need for so much stubbing.

Reply:
> The client code is not publishing those events either, it's expecting them to be published by the instanceTracker call

I'm not sure I understand. If I'm not mistaken all callers to the instance tracker listen on the event bus. These callers should not care where the event comes from.

> Judging by the argument list needed to instantiate one, it seems doable. I've done that in several areas and it runs fast enough, and removes the need for so much stubbing.

+1 on that.